[FEATURE] Implement `on` modifier.

rwjblue opened the pull request:
See https://emberjs.github.io/rfcs/0471-on-modifier.html for details.

Special thanks to @buschtoens's work on [ember-on-modifier](https://github.com/buschtoens/ember-on-modifier) which heavily inspired this implementation.

Commit:
[FEATURE] Implement `on` modifier.

See https://emberjs.github.io/rfcs/0471-on-modifier.html for details.

Special thanks to [buschtoens](https://github.com/buschtoens)'s work on
[ember-on-modifier](https://github.com/buschtoens/ember-on-modifier)
which heavily inspired this implementation.

pzuraq commented on packages/@ember/-internals/glimmer/tests/integration/modifiers/on-test.js:
```diff
@@ -0,0 +1,324 @@
+import { isChrome, isFirefox } from '@ember/-internals/browser-environment';
+import { privatize as P } from '@ember/-internals/container';
+
+const isIE11 = !window.ActiveXObject && 'ActiveXObject' in window;
```
lol

rwjblue replied:
😭 indeed

pzuraq commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+  callback: EventListener,
+  options?: AddEventListenerOptions
+): void {
+  adds++;
```
Assuming these aren't in DEBUGs because you wanted to run these tests against IE?

rwjblue replied:
Yes, exactly. We can ultimately remove if we care, but it's very very cheap and almost certainly won't matter.

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    if (this.shouldUpdate) {
+      let callback = (this.callback = function(this: Element, event) {
```
Do you think, it'd be a bit faster to only do this, if `DEBUG && passive || !SUPPORTS_EVENT_OPTIONS && once`?

So kinda like:

```ts
if (DEBUG && passive || !SUPPORTS_EVENT_OPTIONS && once) {
  let callback = (this.callback = function(this: Element, event) {
    if (DEBUG && passive) {
      event.preventDefault = () => {
        assert(
          `You marked this listener as 'passive', meaning that you must not call 'event.preventDefault()': \n\n${userProvidedCallback}`
        );
      };
    }

    if (!SUPPORTS_EVENT_OPTIONS && once) {
      removeEventListener(this, eventName, callback, options);
    }
    return userProvidedCallback.call(null, event);
  });
} else {
  this.callback = userProvidedCallback;
}
```

Or can we neglect this?

rwjblue replied:
I was thinking that we didn't care, but honestly its a small change and reduces some overhead so seems good to do. Will update.

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+    // used only in the following cases:
+    //
+    // * where there is no options
+    // * `{ once: true | false, passive: true, capture: false }
```
```suggestion
    // * `{ once: true | false, passive: true | false, capture: false }
```

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+    // used only in the following cases:
+    //
+    // * where there is no options
+    // * `{ once: true | false, passive: true, capture: false }
```
```suggestion
    // * `{ once: true | false, passive: true | false, capture: false }
```

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+  updateFromArgs() {
+    let { args } = this;
+
+    let { once, passive, capture }: AddEventListenerOptions = args.named.value();
```
In `ember-on-modifier`, we assert that only these three options are passed to guard against typos. Do you think we should add this here as well?

rwjblue replied:
Hmm, the RFC specifically states the opposite (that passing other things is "ignored"):

From [here](https://emberjs.github.io/rfcs/0471-on-modifier.html#named-parameters):

> The following named parameters will be accepted as options to {{on}}:
> 
> * capture
> * once
> * passive
> 
> These will be passed forward to addEventListener as options in modern browsers, and will be polyfilled in older browsers. Other options will be ignored.
> 

I don't feel super strongly so I just followed the RFC on this.

@pzuraq - Do you think we should change to assert?

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+    addEventListener(state.element, state.eventName, state.callback, state.options);
+
+    state.shouldUpdate = false;
+  }
```
If I'm reading this correctly, once the modifier is installed, it will only ever update to a new listener, but will never be uninstalled. I can't think of a scenario, where the modifier would need to be uninstalled from an element that _does not_ get removed from the DOM afterwards, but I think @chancancode made a case for this once on Discord.

chancancode replied:
👉 https://twitter.com/chancancode/status/1109502949972074496

rwjblue replied:
@buschtoens - We definitely cleanup (see `getDestructor` just below, and the `destroy` method on the `OnModifierState` object. I will also add a test (using the assert count methodology) to ensure we don't accidentally regress this.

buschtoens replied:
Ah, I see! So `getDestructor` returns a `Destroyable`, which is implemented through `destroy()` in `OnModifierState`. I was not familiar with that part of the Glimmer API yet. 👍

rwjblue replied:
Ya, this is one way the "internal" modifiers differ from the public API. The public API _effectively_ does the same thing (the `destroy` method on the state bucket invokes the `destroyModifier` hook on the custom manager).

buschtoens commented on packages/@ember/-internals/glimmer/tests/integration/modifiers/on-test.js:
```diff
@@ -0,0 +1,324 @@
+        if (isChrome || isFirefox) {
+          assert.strictEqual(SUPPORTS_EVENT_OPTIONS, true, 'is true in chrome and firefox');
+        } else if (isIE11) {
+          assert.strictEqual(SUPPORTS_EVENT_OPTIONS, false, 'is true in chrome and firefox');
```
```suggestion
          assert.strictEqual(SUPPORTS_EVENT_OPTIONS, false, 'is false in IE11');
```

Or was this intentional to keep the logout output the same?

rwjblue replied:
Nope, was a copy/paste mistake

buschtoens commented on packages/@ember/-internals/glimmer/tests/integration/modifiers/on-test.js:
```diff
@@ -0,0 +1,324 @@
+
+        runTask(() => this.$('.inner').click());
+
+        assert.verifySteps(['inner clicked', 'outer clicked'], 'uses capture: false by default');
```
I did not know about the [`verifySteps`](https://api.qunitjs.com/assert/verifySteps) API. Nice! 👏

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,241 @@
+
+    assert(
+      'You must pass a valid DOM event name as the first argument to the `on` modifier',
+      args.positional.at(0) !== undefined && typeof args.positional.at(0).value() === 'string'
```
Similar to asserting that only `once`, `passive` and `capture` are passed as named arguments, we could also check that only 2 positional arguments are passed.

buschtoens reviewed (approved):
Looks awesome! Thank you for getting this into core so swiftly. ⚡️

Commit:
Incorporate code review feedback from buschtoens.

* Extract shared utility for a `this` context that errors when used.
* Assert when more positional arguments are provided.
* Avoid wrapping callback needlessly (makes debugging a bit easier,
  since the event handler function is actually the user-land function in
  most cases.
* Incorporates a system to error if any `this` properties are
  accessed from within the callback in debug builds.
* Fixup comments for addEventListener / removeEventListener logic.
* Fix test assertion message.
* Add test ensuring the modifier is destroyed properly.

rwjblue commented:
FYI - Updated to address review feedback:

* Extract shared utility for a `this` context that errors when used.
* Assert when more positional arguments are provided.
* Avoid wrapping callback needlessly (makes debugging a bit easier,
  since the event handler function is actually the user-land function in
  most cases.
* Incorporates a system to error if any `this` properties are
  accessed from within the callback in debug builds.
* Fixup comments for addEventListener / removeEventListener logic.
* Fix test assertion message.
* Add test ensuring the modifier is destroyed properly.

I think this should be good to go once CI is green, in case anyone wants another crack at a review first...

buschtoens commented on packages/@ember/-internals/glimmer/lib/helpers/fn.ts:
```diff
@@ -1,43 +1,11 @@
-  );
-}
-
+const context = buildUntouchableThis('`fn` helper');
```
![Can't touch this](https://media.giphy.com/media/hFIkt5s7MrzI4/giphy.gif)

NullVoxPopuli replied:
yaaaaas

buschtoens commented on packages/@ember/-internals/glimmer/lib/modifiers/on.ts:
```diff
@@ -0,0 +1,262 @@
+        });
+      } else if (DEBUG) {
+        // prevent the callback from being bound to the element
+        this.callback = userProvidedCallback.bind(untouchableContext);
```
Nice! 👍

CvX commented:
@rwjblue The current implementation of `on` modifier isn't FastBoot ready, is it? It throws "TypeError: element.removeEventListener is not a function" where `{{action}}` didn't.

rwjblue commented:
@CvX - That is being tracked over in https://github.com/emberjs/ember.js/issues/17949.
